### PR TITLE
Integration of Parse cloud config for test/module for px-backup to run on the different cloud providers

### PR DESCRIPTION
**Pull request by sn-px**
**Closes the ticket**
https://portworx.atlassian.net/browse/PA-1328

**Special notes for your reviewer**:

cloud config file will be populated during the run time in jenkins job reading from jenkins keystore

latest log : https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/test-creds-job/19/consoleFull
https://jenkins.pwx.dev.purestorage.com/job/Users/job/santoshn/job/test-creds-job/21/consoleFull

### CODE REVIEW

File: drivers/backup/cloud_config.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        "name": "SERVER_NAME",
+        "ip": "SERVER_IP",
+        "export_path": "EXPORT_PATH",
+        "sub_path": "SUB_PATH",
+        "mount_options": "MOUNT_OPTIONS",
+        "encryption_key": "ENCRYPTION_KEY"
+      }
+    }
+  }
+}
```
Comment (snigdha-px) [inline drivers/backup/cloud_config.json:57]: Details for Rancher is missing

File: tests/backup/parse_cloud_config.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	EncryptionKey string `json:"encryption_key"`
+	Tag           string `json:"tag"`
+}
+
+type Config struct {
+	CloudProviders CloudProviders `json:"cloudProviders"`
+	BackupTargets  BackupTargets  `json:"backupTargets"`
+}
+
+func getConfigObj() (*Config, error) {
```
Comment (snigdha-px) [inline tests/backup/parse_cloud_config.go:100]: Please add the description for this function just above the function

File: tests/backup/parse_cloud_config.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	_, err := os.Getwd()
+	// Read JSON file into a variable
+
+	testConfigPath := "../drivers/backup/cloud_config.json"
+	jsonData, err := ioutil.ReadFile(testConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read the test configutation file in the path %s", testConfigPath)
+	}
+	// Parse JSON into Configuration struct
+	var config Config
```
Comment (snigdha-px) [inline tests/backup/parse_cloud_config.go:111]: Can we please move this variable declaration at the start of the function

**Review by snigdha-px — CHANGES_REQUESTED** [review]

**Review by mkoppal-px — CHANGES_REQUESTED** [review]
- Both the files `cloud_config.json` and `parse_cloud_config.go` should be kept in a separate package. They don't fit in `backup` or `tests` package.
- I think `cloud_creds.json` and `cloud_creds.go` would be a more apt name
- At what stage will you populate the json file with the actual creds and at what stage will you call `getConfigObj`?
- If you are planning to call `getConfigObj` in every test case then we need to rethink because I was hoping that all the creds are available as an environment variable. Unmarshalling the json in every test case is not a good design.

File: drivers/backup/cloud_config.json [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        "region": "AWS_REGION"
+      }
+    },
+    "azure": {
+      "default": {
+        "subscription_id": "AZURE_SUBSCRIPTION_ID",
+        "client_id": "AZURE_CLIENT_ID",
+        "client_secret": "AZURE_CLIENT_SECRET",
+        "tenant_id": "AZURE_TENANT_ID"
+      }
```
Comment (snigdha-px) [inline drivers/backup/cloud_config.json:16]: Please once check the method : GetAzureCredsFromEnv
AZURE_ACCOUNT_NAME and AZURE_ACCOUNT_KEY also needs to be initialized here.

**Review by snigdha-px — CHANGES_REQUESTED** [review]

**Comment by mkoppal-px** [comment]
- Add doc strings to all the new functions
- Make the parsed object global
- Make the parsing happen in before suite

File: drivers/backup/cloud_config.json (hunk 1 above)
Comment (sn-px) [reply drivers/backup/cloud_config.json:59]: done added

File: tests/backup/parse_cloud_config.go (hunk 3 above)
Comment (sn-px) [reply tests/backup/parse_cloud_config.go]: done added

File: drivers/backup/cloud_config.json (hunk 4 above)
Comment (sn-px) [reply drivers/backup/cloud_config.json:18]: done

File: tests/backup/parse_cloud_config.go (hunk 2 above)
Comment (sn-px) [reply tests/backup/parse_cloud_config.go]: done

File: tests/backup/backup_test.go [hunk 5]
```diff
@@ -34,6 +34,18 @@ var _ = Describe("{BackupClusterVerification}", func() {
 	})
 })
 
+var _ = Describe("{testParseConfig}", func() {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:37]: Can we remove this before merging?

File: tests/backup/backup_basic_test.go [hunk 6]
```diff
@@ -137,6 +139,7 @@ var _ = BeforeSuite(func() {
 	StartTorpedoTest("Setup buckets", "Creating one generic bucket to be used in all cases", nil, 0)
 	defer EndTorpedoTest()
 	// Create the first bucket from the list to be used as generic bucket
+	globalConfig, _ = backup.GetConfigObj()
```
Comment (mkoppal-px) [inline tests/backup/backup_basic_test.go:142]: Catch the error here

File: drivers/backup/parse_cloud_config.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+	CloudProviders CloudProviders `json:"cloudProviders"`
+	BackupTargets  BackupTargets  `json:"backupTargets"`
+}
+
+// GetConfigObj reads the configuration file and returns a Config object.
+func GetConfigObj() (*Config, error) {
+	var config Config
+	_, err := os.Getwd()
+	// Read JSON file into a variable
+	testConfigPath := "../drivers/backup/cloud_config.json"
```
Comment (mkoppal-px) [inline drivers/backup/parse_cloud_config.go:95]: Create a PA ticket to make this path configurable

File: tests/backup/backup_test.go (hunk 5 above)
Comment (sn-px) [reply tests/backup/backup_test.go:37]: discussed with mithun will add a comment to keep it as reference test

File: tests/backup/backup_basic_test.go [hunk 8]
```diff
@@ -130,12 +132,16 @@ func BackupInitInstance() {
...
 var _ = BeforeSuite(func() {
+	var err error
 	dash = Inst().Dash
 	dash.TestSetBegin(dash.TestSet)
 	log.Infof("Backup Init instance")
 	BackupInitInstance()
 	StartTorpedoTest("Setup buckets", "Creating one generic bucket to be used in all cases", nil, 0)
 	defer EndTorpedoTest()
+	// Get all the values from the cloud_config.json persist into struct which can be globally accessed
+	globalConfig, err = backup.GetConfigObj()
```
Comment (mkoppal-px) [inline tests/backup/backup_basic_test.go:143]: Rename this to `GlobalCredentialConfig`

File: drivers/backup/parse_cloud_config.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
+package backup
```
Comment (mkoppal-px) [inline drivers/backup/parse_cloud_config.go:1]: Rename this file to `cloud_cred_config.go`

File: tests/backup/backup_basic_test.go (hunk 6 above)
Comment (sn-px) [reply tests/backup/backup_basic_test.go]: done

File: tests/backup/backup_basic_test.go (hunk 8 above)
Comment (sn-px) [reply tests/backup/backup_basic_test.go:143]: done

File: drivers/backup/parse_cloud_config.go (hunk 9 above)
Comment (sn-px) [reply drivers/backup/parse_cloud_config.go]: done

**Comment by sn-px** [comment]
> * Both the files `cloud_config.json` and `parse_cloud_config.go` should be kept in a separate package. They don't fit in `backup` or `tests` package.
> * I think `cloud_creds.json` and `cloud_creds.go` would be a more apt name
> * At what stage will you populate the json file with the actual creds and at what stage will you call `getConfigObj`?
> * If you are planning to call `getConfigObj` in every test case then we need to rethink because I was hoping that all the creds are available as an environment variable. Unmarshalling the json in every test case is not a good design.

done

**Comment by sn-px** [comment]
> * Add doc strings to all the new functions
> * Make the parsed object global
> * Make the parsing happen in before suite

done

**Review by snigdha-px — APPROVED** [review]
LGTM

**Pull request merged**